packages/user-extension/src/index.ts

```diff
@@ -0,0 +1,79 @@
+// Copyright (c) Jupyter Development Team.
+// Distributed under the terms of the Modified BSD License.
+/**
+ * @packageDocumentation
+ * @module user-extension
+ */
+
+import {
+  JupyterFrontEnd,
+  JupyterFrontEndPlugin
+} from '@jupyterlab/application';
+import { PageConfig } from '@jupyterlab/coreutils';
+import { IStateDB } from '@jupyterlab/statedb';
+import {
+  ICurrentUser,
+  IMenu,
+  IUser,
+  IUserMenu,
+  RendererUserMenu,
+  User,
+  UserMenu
+} from '@jupyterlab/user';
+import { MenuBar, Widget } from '@lumino/widgets';
+
+/**
+ * Jupyter plugin providing the ICurrentUser.
+ */
+const userPlugin: JupyterFrontEndPlugin<IUser> = {
+  id: '@jupyterlab/user-extension:user',
+  autoStart: true,
+  requires: [IStateDB],
+  provides: ICurrentUser,
+  activate: (app: JupyterFrontEnd, state: IStateDB): IUser => {
+    return new User(state);
+  }
+};
+
+/**
+ * Jupyter plugin providing the IUserMenu.
+ */
+const userMenuPlugin: JupyterFrontEndPlugin<IMenu | undefined> = {
```

Comment:
Two questions here:

- for clarity and coherence with the rest of the code base, should this be providing a `IUserMenu` instead of `IMenu`?
- probably the plugin should always provide `IUserMenu` instead of `IMenu | undefined`. Then another plugin would be responsible for adding it to the shell by checking the `collaborative` option. This would be similar to the pattern used for enabling the debugger sidebar here:

https://github.com/jupyterlab/jupyterlab/blob/1bf5fdff171f8a9e91755821efd376d5742778ba/packages/debugger-extension/src/index.ts#L565-L582

Reply:
> for clarity and coherence with the rest of the code base, should this be providing a IUserMenu instead of IMenu

I'm lost in this part.
The `IUserMenu` is the name of the token that the plugin provides, And the `IMenu` is the interface of the object provided by the plugin.
So, the plugin provides an `IUserMenu` and returns an `IMenu`, right?

> probably the plugin should always provide IUserMenu instead of IMenu | undefined. Then another plugin would be responsible for adding it to the shell by checking the collaborative option. This would be similar to the pattern used for enabling the debugger sidebar here:
Thanks for the clarification. I had no idea how to proceed in this case.